Use correct stream for output.

bangerth opened the pull request:
Many of the autodifferentiation tests output parts to `deallog` and other parts of their output to `std::cout`. The latter is not captured in the `.output` files and consequently not checked. I think that this is accidental.

Addresses part of #16534, though there is more to do.

tamiko reviewed (approved):
Correct the current logic is that the output of `deallog` will be used if output is present. If nothing is printed to `deallog` then everything piped to `std::cout` is used for comparison instead.

If I remember correctly, it was not possible to combine both, i.e., using `deallog` and `std::cout` together for comparison, back when we introduced that behavior because that lead to quite a number of failing tests.

bangerth commented:
Yes, not worth trying to support outputting to both.

masterleinad commented on tests/symengine/symengine_tensor_operations_03.cc:
```diff
@@ -507,37 +507,37 @@ test_symmetric_tensor_tensor_vector_scalar_coupled(
-  print(std::cout, "symb_d2psi_ds_x_dst", symb_d2psi_ds_x_dst);
-  print(std::cout, "symb_d2psi_ds_x_dt", symb_d2psi_ds_x_dt);
-  print(std::cout, "symb_d2psi_ds_x_dv", symb_d2psi_ds_x_dv);
-  print(std::cout, "symb_d2psi_ds_x_ds", symb_d2psi_ds_x_ds);
+  print(deallog, "symb_st", symb_st);
+  print(deallog, "symb_t", symb_t);
+  print(deallog, "symb_v", symb_v);
+  print(deallog, "symb_s", symb_s);
+
+  print(deallog, "symb_psi", symb_psi);
+
+  print(deallog, "symb_dpsi_dst", symb_dpsi_dst);
+  print(deallog, "symb_dpsi_dt", symb_dpsi_dt);
+  print(deallog, "symb_dpsi_dv", symb_dpsi_dv);
+  print(deallog, "symb_dpsi_ds", symb_dpsi_ds);
+
+  print(deallog, "symb_d2psi_dst_x_dst", symb_d2psi_dst_x_dst);
+  print(deallog, "symb_d2psi_dst_x_dt", symb_d2psi_dst_x_dt);
+  print(deallog, "symb_d2psi_dst_x_dv", symb_d2psi_dst_x_dv);
+  print(deallog, "symb_d2psi_dst_x_ds", symb_d2psi_dst_x_ds);
+
+  print(deallog, "symb_d2psi_dt_x_dst", symb_d2psi_dt_x_dst);
+  print(deallog, "symb_d2psi_dt_x_dt", symb_d2psi_dt_x_dt);
+  print(deallog, "symb_d2psi_dt_x_dv", symb_d2psi_dt_x_dv);
+  print(deallog, "symb_d2psi_dt_x_ds", symb_d2psi_dt_x_ds);
+
+  print(deallog, "symb_d2psi_dv_x_dst", symb_d2psi_dv_x_dst);
+  print(deallog, "symb_d2psi_dv_x_dt", symb_d2psi_dv_x_dt);
+  print(deallog, "symb_d2psi_dv_x_dv", symb_d2psi_dv_x_dv);
+  print(deallog, "symb_d2psi_dv_x_ds", symb_d2psi_dv_x_ds);
+
+  print(deallog, "symb_d2psi_ds_x_dst", symb_d2psi_ds_x_dst);
+  print(deallog, "symb_d2psi_ds_x_dt", symb_d2psi_ds_x_dt);
+  print(deallog, "symb_d2psi_ds_x_dv", symb_d2psi_ds_x_dv);
+  print(deallog, "symb_d2psi_ds_x_ds", symb_d2psi_ds_x_ds);
 #endif
```
This is guarded by `DEBUG` and so you will get different output in `Debug` and `Release` mode but there is only one output file AFAICT.

bangerth commented:
I fixed the issue @masterleinad points out by always outputting. I also fixed a few other places where we were writing to `std::cout`. This required restricting output operators in the library for some custom enum types to `std::ostream` from a general template, because the general template was ambiguous to `operator<< (LogStream &, T &)`. I put that into the very first commit. These output operators are sufficiently obscure and almost certainly not widely used that I didn't care too much about the compatibility impact.

masterleinad commented on tests/symengine/step-44-sd-quadrature_level_01.cc:
```diff
@@ -97,7 +97,7 @@ namespace Step44
       if (debug)
         {
-          SD::Utilities::print_substitution_map(std::cout, sub_vals);
+          SD::Utilities::print_substitution_map(deallog, sub_vals);
```
The output here depends on a bool variable that is `false`. If someone was to change that, the test would then fail. Ultimately, it seems that this output is not intended to be recorded in an output file but only printed to the screen for debugging purposes.

bangerth replied:
True. There are other places where these programs have `#ifdef DEBUG_EXTRA`. It is true that one could address all of these things, but that's more work than I'm willing to put in at the moment. I'm just trying to make things better, not optimal.

masterleinad replied:
Fair enough.
